hikari/interactions/modal_interactions.py

```diff
@@ -0,0 +1,359 @@
+
+    def __iter__(self) -> typing.Iterator[PartialModalComponent]:
+        return iter(self.components)
+
+    def __len__(self) -> int:
+        return len(self.components)
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class InteractionTextInput(PartialModalComponent):
+    """A text input component in a modal interaction."""
+
+    custom_id: str = attr.field(repr=True)
+    """Developer set custom ID used for identifying interactions with this modal."""
+
+    value: str = attr.field(repr=True)
+    """Value provided for this text input."""
+
+
+@attr_extensions.with_copy
+@attr.define(hash=True, kw_only=True, weakref_slot=False)
+class ModalInteraction(base_interactions.MessageResponseMixin[ModalResponseTypesT]):
+    """Represents a modal interaction on Discord."""
+
+    channel_id: snowflakes.Snowflake = attr.field(eq=False, hash=False, repr=True)
+    """ID of the channel this modal interaction event was triggered in."""
+
+    custom_id: str = attr.field(eq=False, hash=False, repr=True)
+    """The custom id of the modal."""
+
+    guild_id: typing.Optional[snowflakes.Snowflake] = attr.field(eq=False, hash=False, repr=True)
+    """ID of the guild this modal interaction event was triggered in.
+
+    This will be `builtins.None` for modal interactions triggered in DMs.
+    """
+
+    guild_locale: typing.Optional[str] = attr.field(eq=False, hash=False, repr=True)
+    """The preferred language of the guild this modal interaction was triggered in.
+
+    This will be `builtins.None` for modal interactions triggered in DMs.
+
+    !!! note
+        This value can usually only be changed if `COMMUNITY` is in `hikari.guilds.Guild.features`
+        for the guild and will otherwise default to `en-US`.
+    """
+
+    message: typing.Optional[messages.Message] = attr.field(eq=False, repr=False)
+    """The message whose component triggered the modal.
+
+    This will be None if the modal was a response to a command.
+    """
+
+    member: typing.Optional[base_interactions.InteractionMember] = attr.field(eq=False, hash=False, repr=True)
+    """The member who triggered this modal interaction.
+
+    This will be `builtins.None` for modal interactions triggered in DMs.
+
+    !!! note
+        This member object comes with the extra field `permissions` which
+        contains the member's permissions in the current channel.
+    """
+
+    user: _users.User = attr.field(eq=False, hash=False, repr=True)
+    """The user who triggered this modal interaction."""
+
+    locale: str = attr.field(eq=False, hash=False, repr=True)
+    """The selected language of the user who triggered this modal interaction."""
+
+    app_permissions: typing.Optional[permissions.Permissions] = attr.field(eq=False, hash=False, repr=False)
+    """Permissions the bot has in this interaction's channel if it's in a guild."""
+
+    components: typing.Sequence[PartialModalComponent] = attr.field(eq=False, hash=False, repr=True)
```

Comment:
This field will actually hold the rows of type `ModalActionRowComponent `right?

Why not having it typed as `typing.Sequence[ModalActionRowComponent]` then?

Reply:
This will not always be correct, as discord might add more types in the future. This makes it more "future-proof" and provides a better DX

Reply:
Whatever types are added in the future, there will always be wrapped in an ActionRow no?

Reply:
Could be, could also be that discord add a separate type. Better safe than sorry

Reply:
I understand the point, but in the meantime we have to assert types everytime for no big reason, especially for modals: you want to retrieve what the user filled so you need to iterate through all the fields.

```py
for row in interaction.components:
    # Have to check row is an Action row
    for comp in row.components:
        # Have to check the type of component (makes sense here)
        foo(comp)
```

Reply:
Changed :)